src/remap.cpp

```diff
@@ -294,33 +291,112 @@ void Remap::run(int particle, bool accumulate, bool rescale, double scaling) {
-    auto const range = HostRange(0, remapped_values.size());
+    auto slice = mesh.get_field_slice(mesh.fields, i);
+    auto const& values = target.get_field(fields[i]);
 
     if (accumulate) {
-      Kokkos::parallel_for(range, [&](int j) { field(j) += remapped_values[j]; });
+      Kokkos::parallel_for(range, [&](int j) { slice(j) += values[j]; });
     } else {
-      Kokkos::parallel_for(range, [&](int j) { field(j) = remapped_values[j]; });
+      Kokkos::parallel_for(range, [&](int j) { slice(j) = values[j]; });
     }
     if (rescale) {
-      Kokkos::parallel_for(range, [&](int j) { field(j) *= scaling; });
+      Kokkos::parallel_for(range, [&](int j) { slice(j) *= scaling; });
     }
   }
 }
 
+/* -------------------------------------------------------------------------- */
+void Remap::estimate_gradients() {
+
+  if (input.mpi.rank != 0) { return; }
+
+  std::cout << "estimate gradients ... " << std::flush;
+
+  // step 0: update mesh points coordinates
+  update_mesh(false);
+
+  Wonton::vector<Wonton::Vector<DIM>> gradients(mesh.num_points);
+  Wonton::vector<std::vector<Point<DIM>>> coordinates(mesh.num_points);
+  Wonton::vector<std::vector<Wonton::Matrix>> stencils(mesh.num_points);
+  Wonton::vector<std::vector<double>> values(mesh.num_points);
+
+  // step 1: retrieve the neighbors of each mesh point
+  using Filter = Portage::SearchPointsBins<DIM, Wonton::Swarm<DIM>, Wonton::Swarm<DIM>>;
+
+  Filter search(grid, grid, extents, extents, WeightCenter::Gather, 3);
+  Wonton::transform(grid.begin(Wonton::PARTICLE, Wonton::PARALLEL_OWNED),
+                    grid.end(Wonton::PARTICLE, Wonton::PARALLEL_OWNED),
+                    neighbors.begin(), search);
+
+  // enrich each list of neighbors by prepending it with the current point index
+  Wonton::for_each(grid.begin(Wonton::PARTICLE, Wonton::PARALLEL_OWNED),
+                   grid.end(Wonton::PARTICLE, Wonton::PARALLEL_OWNED),
+                   [&](int i) {
+                     std::vector<int> indices = neighbors[i];
+                     indices.emplace(indices.begin(), i);
+                     neighbors[i] = indices;
+                   });
+
+
+  // retrieve the coordinates of each point of the stencil
+  Wonton::transform(neighbors.begin(), neighbors.end(), coordinates.begin(),
+                    [&](auto const& indices) {
+                      std::vector<Wonton::Point<DIM>> points;
+                      for (int j : indices) {
+                        points.emplace_back(grid.get_particle_coordinates(j));
+                      }
+                      return points;
+                    });
+
+
+  // step 2: build and cache stencil matrices
+  Wonton::transform(coordinates.begin(), coordinates.end(), stencils.begin(),
+                    [&](auto const& points) {
+                      return Wonton::build_gradient_stencil_matrices<DIM>(points, true);
+                    });
+
+  for (int f = 0; f < num_fields; ++f) {
```

Comment:
Finer control of which field to take the derivative will be useful here. Right now, only the last field is a potential, and we need only the longitudinal derivative (along x'). It is probably harder to just take derivative in one direction, so we can leave that as it is(will need both directions in the case of full potential kernel).

Reply:
A filtering can be applied to only take the derivative of a given field in a particular direction. Such an option can then be exposed to the user within the input deck. But given how little time it takes to compute the gradients, I don't know if it would be worthwhile.
Actually, if we really want to reduce the computational cost, we have to update the matrix A to take scalars instead of 2D coordinates, as well as the data structures to take scalar elements instead of points. But if there are cases where we want to take both derivatives for a given field, then reworking the code would be pointless.